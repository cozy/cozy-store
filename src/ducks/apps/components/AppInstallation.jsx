import React, { Component } from 'react'

import { connect } from 'react-redux'
import { PropTypes } from 'react-proptypes'
import flags from 'cozy-flags'

import ReactMarkdownWrapper from 'ducks/components/ReactMarkdownWrapper'
import { ModalDescription, ModalHeader, ModalFooter } from 'cozy-ui/react/Modal'
import Spinner from 'cozy-ui/react/Spinner'
import Button from 'cozy-ui/react/Button'

import PermissionsList from 'ducks/apps/components/PermissionsList'
import Partnership from 'ducks/apps/components/Partnership'
import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'
import Checkbox from 'cozy-ui/react/Checkbox'

import { getTranslatedManifestProperty } from 'lib/helpers'
import { hasPendingUpdate } from 'ducks/apps/appStatus'
import storeConfig from 'config'

import { APP_TYPE, getAppBySlug, installAppFromRegistry } from 'ducks/apps'

const shouldSkipPermissions = app =>
  flags('skip-low-permissions') &&
  app.label <= storeConfig.default.authorizedLabelLimit

const hasInstallation = app => !app.installed || hasPendingUpdate(app)

export class AppInstallation extends Component {
  constructor(props) {
    super(props)
    const { app } = props
    this.state = { isTermsAccepted: false }
    this.installApp = this.installApp.bind(this)
    if (shouldSkipPermissions(app) && hasInstallation(app)) {
      this.installApp()
    }
  }

  installApp() {
    if (!this.isInstallReady()) return // Not ready to be installed
    const { app, channel, installApp, onInstallOrUpdate, t } = this.props
    const isUpdate = app.installed
    const isChannelSwitch = !!channel
    installApp(app, channel, isUpdate).catch(error => {
      const appName = getTranslatedManifestProperty(app, 'name', t)
      Alerter.error(
        t('app_modal.install.alert.install_error', {
          message: error.message,
          name: appName
        }),
        {
          duration: 12000,
          buttonText: t('app_modal.install.alert.dismiss'),
          buttonAction: dismiss => dismiss()
        }
      )
      console.error(error)
    })
    if (isUpdate || isChannelSwitch || app.type !== APP_TYPE.KONNECTOR) {
      if (typeof onInstallOrUpdate === 'function') onInstallOrUpdate()
    }
  }

  componentDidUpdate = prevProps => {
    const { app, onKonnectorInstall } = this.props
    const justInstalled =
      !!prevProps.app && !prevProps.app.installed && app.installed
    if (justInstalled && app.type === APP_TYPE.KONNECTOR) {
      if (typeof onKonnectorInstall === 'function') onKonnectorInstall()
    }
  }

  acceptTerms = () => {
    this.setState(state => ({
      isTermsAccepted: !state.isTermsAccepted
    }))
  }

  isInstallReady = () => {
    const { app, isInstalling } = this.props
    const isCurrentAppInstalling = isInstalling === app.slug
    const isTermsReady =
      shouldSkipPermissions(app) || (!app.terms || this.state.isTermsAccepted)
    return !isCurrentAppInstalling && isTermsReady
  }

  render() {
    const {
      app,
      fetchError,
      installError,
      isFetching,
      isAppFetching,
      isInstalling,
      onCancel,
      t
    } = this.props
    const { isTermsAccepted } = this.state
    const isFetchingSomething = isFetching || isAppFetching
    const isCurrentAppInstalling = isInstalling === app.slug

    if (shouldSkipPermissions(app) && hasInstallation(app)) {
      return (
        <React.Fragment>
          <ModalDescription className="sto-install-loading-wrapper">
            <div className="sto-install-loading">
              <Spinner size="xxlarge" />
            </div>
          </ModalDescription>
        </React.Fragment>
      )
    }

    return (
      <React.Fragment>
        <ModalHeader title={t('app_modal.install.title')} />
        {isFetchingSomething ? (
          <ModalDescription className="sto-install-loading-wrapper">
            <div className="sto-install-loading">
              <Spinner size="xlarge" />
            </div>
          </ModalDescription>
        ) : (
          <ModalDescription>
            {app.partnership && <Partnership app={app} />}
            {app.permissions && <PermissionsList app={app} />}
            {fetchError && (
              <p className="u-error">
                {t('app_modal.install.message.version_error', {
                  message: fetchError.message
                })}
              </p>
            )}
          </ModalDescription>
        )}
        {!isFetchingSomething && !fetchError && (
          <ModalFooter className="sto-install-footer">
            {installError && (
              <p className="u-error">
                {t('app_modal.install.message.install_error', {
                  message: installError.message
                })}
              </p>
            )}
            {app.terms && (
              <div className="sto-install-terms">
                <Checkbox
                  className="sto-install-terms-checkbox"
                  onChange={this.acceptTerms}
                  checked={isTermsAccepted}
                  disabled={isInstalling}
                >
                  <ReactMarkdownWrapper
                    source={t('app_modal.install.terms', {
                      url: app.terms.url
                    })}
                  />
                </Checkbox>
              </div>
            )}
            <div className="sto-install-controls">
              <Button
                theme="secondary"
                onClick={onCancel}
                disabled={isCurrentAppInstalling}
                label={t('app_modal.install.cancel')}
                extension="full"
                className="u-mh-half"
              />
              <Button
                theme="primary"
                disabled={!this.isInstallReady() || isInstalling}
                busy={isCurrentAppInstalling}
                extension="full"
                onClick={this.installApp}
                label={
                  hasPendingUpdate(app)
                    ? t('app_modal.install.update')
                    : t('app_modal.install.install')
                }
                className="u-mh-half"
              />
            </div>
          </ModalFooter>
        )}
      </React.Fragment>
    )
  }
}

AppInstallation.propTypes = {
  appSlug: PropTypes.string.isRequired
}

const mapStateToProps = (state, ownProps) => ({
  app: getAppBySlug(state, ownProps.appSlug),
  isFetching: state.apps.isFetching,
  isAppFetching: state.apps.isAppFetching,
  isInstalling: state.apps.isInstalling,
  fetchError: state.apps.fetchError,
  installError: state.apps.actionError
})

const mapDispatchToProps = dispatch => ({
  installApp: (app, channel, isUpdate) =>
    dispatch(installAppFromRegistry(app, channel, isUpdate))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(AppInstallation))
