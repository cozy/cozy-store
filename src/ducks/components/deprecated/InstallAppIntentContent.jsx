import React from 'react'

import IntentHeader from 'cozy-ui/transpiled/react/IntentHeader'
import Spinner from 'cozy-ui/transpiled/react/Spinner'

import AppInstallationDeprecated from 'ducks/apps/components/deprecated/AppInstallation'
import InstallSuccessDeprecated from 'ducks/apps/components/deprecated/InstallSuccess'

/**
 * @deprecated
 */
const InstallAppIntentContent = ({
  appData,
  fetching,
  error,
  isReadyWithoutErrors,
  isInstalled,
  app,
  isInstalling,
  onCancel,
  onTerminate,
  installApp,
  hasWebappSucceed
}) => {
  return (
    <div className="coz-intent-wrapper">
      <IntentHeader
        appEditor={appData.app.editor}
        appName={appData.app.name}
        appIcon={`../${appData.app.icon}`}
      />
      <div className={`coz-intent-content${fetching ? ' --loading' : ''}`}>
        {fetching && <Spinner size="xxlarge" noMargin />}
        {error && <div className="coz-error">{error.message}</div>}
        {isReadyWithoutErrors &&
          (!isInstalled ? (
            <AppInstallationDeprecated
              appSlug={app.slug}
              installApp={installApp}
              isInstalling={isInstalling}
              onCancel={onCancel}
            />
          ) : (
            hasWebappSucceed && (
              <InstallSuccessDeprecated app={app} onTerminate={onTerminate} />
            )
          ))}
      </div>
    </div>
  )
}

export default InstallAppIntentContent
