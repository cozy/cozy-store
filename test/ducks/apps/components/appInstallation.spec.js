import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'

import flags from 'cozy-flags'

import { tMock } from '../../../jestLib/I18n'
import { AppInstallation } from 'ducks/apps/components/deprecated/AppInstallation'
import storeConfig from 'config'

import mockRegistryApp from '../_mockPhotosRegistryVersion'

describe('AppInstallation component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
    jest.spyOn(AppInstallation.prototype, 'installApp')
  })
  afterEach(() => {
    AppInstallation.prototype.installApp.mockRestore()
  })

  it('should install automatically the app if the label is skippable and render the loading modal', () => {
    flags('skip-low-permissions', true)
    const mockApp = Object.assign({}, mockRegistryApp, {
      label: storeConfig.default.authorizedLabelLimit
    })
    const mockInstallProp = jest.fn(() => Promise.resolve())
    const { container, getByRole } = render(
      <AppInstallation
        t={tMock}
        appSlug={mockApp.slug}
        app={mockApp}
        onCancel={jest.fn()}
        installApp={mockInstallProp}
      />
    )
    expect(mockInstallProp).toHaveBeenCalledTimes(1)
    expect(AppInstallation.prototype.installApp).toHaveBeenCalledTimes(1)
    expect(container.querySelector('.sto-install-loading')).toBeInTheDocument()
    expect(getByRole('progressbar')).toBeInTheDocument()
    flags('skip-low-permissions', false)
  })

  it('should not install automatically the app if the label is not skippable and render the permissions modal', () => {
    flags('skip-low-permissions', true)
    const mockApp = Object.assign({}, mockRegistryApp, {
      label: storeConfig.default.authorizedLabelLimit + 1
    })
    const mockInstallProp = jest.fn(() => Promise.resolve())
    const { container, getByRole } = render(
      <AppInstallation
        t={tMock}
        appSlug={mockApp.slug}
        app={mockApp}
        onCancel={jest.fn()}
        installApp={mockInstallProp}
      />
    )
    expect(mockInstallProp).not.toHaveBeenCalled()
    expect(AppInstallation.prototype.installApp).not.toHaveBeenCalled()
    expect(container.querySelector('.sto-install-controls')).toBeInTheDocument()
    expect(getByRole('button', { name: 'Cancel' })).toBeInTheDocument()
    expect(getByRole('button', { name: 'Install' })).toBeInTheDocument()
    flags('skip-low-permissions', false)
  })
})
