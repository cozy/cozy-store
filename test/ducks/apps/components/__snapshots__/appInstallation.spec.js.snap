// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInstallation component should install automatically the app if the label is skippable and render the loading modal 1`] = `
<React.Fragment>
  <ModalContent
    className="sto-install-loading-wrapper"
  >
    <div
      className="sto-install-loading"
    >
      <withI18n(Spinner)
        size="xxlarge"
      />
    </div>
  </ModalContent>
</React.Fragment>
`;

exports[`AppInstallation component should not install automatically the app if the label is not skippable and render the permissions modal 1`] = `
<React.Fragment>
  <ModalHeader
    title="Data usage"
  />
  <ModalContent />
  <ModalFooter
    className="sto-install-footer"
  >
    <div
      className="sto-install-controls"
    >
      <DefaultButton
        className="u-mh-half"
        disabled={false}
        extension="full"
        label="Cancel"
        onClick={[MockFunction]}
        theme="secondary"
      />
      <DefaultButton
        busy={false}
        className="u-mh-half"
        extension="full"
        label="Install"
        onClick={[Function]}
        theme="primary"
      />
    </div>
  </ModalFooter>
</React.Fragment>
`;
